src/main/scala/com/exasol/cloudetl/kinesis/KinesisShardsMetadataReader.scala

```diff
@@ -0,0 +1,66 @@
+package com.exasol.cloudetl.kinesis
+
+import scala.collection.JavaConverters._
+
+import com.amazonaws.auth._
+import com.amazonaws.services.kinesis.{AmazonKinesis, AmazonKinesisClientBuilder}
+import com.amazonaws.services.kinesis.model._
+import com.exasol._
+import java.util
+
+object KinesisShardsMetadataReader {
+  def run(exaMetadata: ExaMetadata, exaIterator: ExaIterator): Unit = {
+    val kinesisArgumentsReader = new KinesisArgumentsReader(exaIterator)
+    val awsAccessKeyId = kinesisArgumentsReader.getAwsAccessKeyArgument
+    val awsSecretAccessKey = kinesisArgumentsReader.getAwsSecretKeyArgument
+    val awsSessionToken = kinesisArgumentsReader.getAwsSessionTokenArgument
+    val streamName = kinesisArgumentsReader.getStreamNameArgument
+    val region = kinesisArgumentsReader.getRegionArgument
+    val awsCredentials =
+      new BasicSessionCredentials(awsAccessKeyId, awsSecretAccessKey, awsSessionToken)
+    val amazonKinesis = AmazonKinesisClientBuilder.standard
+      .withRegion(region)
+      .withCredentials(new AWSStaticCredentialsProvider(awsCredentials))
+      .build
+    val shards = getAllShardsFromStream(streamName, amazonKinesis)
+    val kinesisMetadata = collectShardsIds(shards)
+    try {
+      kinesisMetadata.foreach(
+        shardId => exaIterator.emit(shardId)
+      )
+    } catch {
+      case exception @ (_: ExaDataTypeException | _: ExaIterationException) =>
+        val message = exception.getMessage
+        throw new KinesisConnectorException(
+          s"KinesisShardsMetadataReader cannot emit shards' ids. Caused: $message",
+          exception
+        )
+    }
+  }
+
+  private[kinesis] def getAllShardsFromStream(
+    streamName: String,
+    amazonKinesis: AmazonKinesis
+  ): List[Shard] = {
+    val describeStreamRequest = new DescribeStreamRequest
+    describeStreamRequest.setStreamName(streamName)
+    val shards = new util.ArrayList[Shard]
+    var exclusiveStartShardId: String = null
+    do {
+      describeStreamRequest.setExclusiveStartShardId(exclusiveStartShardId)
+      val describeStreamResult = amazonKinesis.describeStream(describeStreamRequest)
+      val _ = shards.addAll(describeStreamResult.getStreamDescription.getShards)
+      if (describeStreamResult.getStreamDescription.getHasMoreShards && !shards.isEmpty) {
+        exclusiveStartShardId = shards.get(shards.size - 1).getShardId
+      } else {
+        exclusiveStartShardId = null
+      }
+    } while ({
+      exclusiveStartShardId != null
+    })
```

Comment:
Can the loop be replaced with `describeStreamResult.getStreamDescription.getShards`?

Reply:
I used a code example from the documentation: https://docs.aws.amazon.com/streams/latest/dev/kinesis-using-sdk-java-retrieve-shards.html

It is said that `getShards` can return not everything from the first call, that's why it's in a loop. We check if some shards left until we get every shard.